### ISSUE TITLE
Distributed operations on DataFrame read using engine='arrow' fail with axis 'Length mismatch'

### ISSUE THREAD

**Issue by bendruitt** [dask/dask#2905]
I have a project which generates a large number of parquet files with high dimensionality (100k+ columns). I have found the metadata files generated by the fastparquet library to be too slow to write and read my data files. I have therefore decided to not use the write_parquet method in dask but write the parquet file chunks manually from my map_partitions function using pyarrow.parquet.write_table. This works well and I can read and write the chunks much faster than with fastparquet. My problem occurs when trying to do any distributed operations on these datasets.

For the exmaples below, I have written the simple DataFrame below to 3 pyarrow.parquet files:
`df = pd.DataFrame(data=np.arange(0,50).reshape(10, 5), columns=['col_A', 'col_B', 'col_C', 'col_D', 'col_E'])`
`df.set_index('col_A', inplace=True)`

Basic example 1:
`read_ddf = dd.read_parquet('chunks', engine='arrow')`
`read_ddf.compute()`

=> this works!

Basic example 2:
`read_ddf = dd.read_parquet('chunks', engine='arrow', columns=['col_B'])`
`read_ddf.compute()`

=> returns a DataFrame with col_B and NO index.

Basic example 3:
`read_ddf = dd.read_parquet('chunks', engine='arrow')`
`dropped_ddf = read_ddf.drop('col_B', axis=1)`
`dropped_ddf.compute()`

=> ValueError: Length mismatch: Expected axis has 3 elements, new values have 4 elements

Any distributed function that attempts to use my dataframe read using engine='arrow' fails with similar mismatches in axis length. I believe this is caused by the way in which the index is handled in the original dataframe.

In summary, using engine='arrow' with the dask read_parquet function to read a folder of parquet files which does not include a metadata file does not seem to handle the index properly for subsequent distributed operations and fails.

Hope this helps!

**Comment by xhochy** [comment]
Hello @bendruitt , 

some things on your problems:

- Example 2 should be fixed with the latest git master of `pyarrow`. We had a bug that the index was not loaded if a column selection was passed to the read method. This will come with the next Apache Arrow release. If you cannot wait that long, feel free to build from source or use `conda install pyarrow -c twosigma`
- Example 3 might also be caused by this problem. It would be nice if you could try out the latest `pyarrow` version.
- The newest `dask` master also has `write_parquet` support with `engine='arrow'`. In combination with the `pyarrow` master, this should make your life easier.

**Comment by jcrist** [comment]
> It would be nice if you could try out the latest pyarrow version

Note that until #2901 is fixed, files written using `pyarrow` master will *probably* not be readable using dask. Files written using `0.7.1` should be though.

> The newest dask master also has write_parquet support with engine='arrow'.

We just released dask `0.16.0`, which has support for writing using `pyarrow`, so you shouldn't need to install from master. Note also that in the new release `engine='arrow'` has been deprecated in favor of `engine='pyarrow'` to better match pandas `to_parquet`.

**Comment by jcrist** [comment]
Actually, this is because`pyarrow` also requires a metadata file, just one with less information than `fastparquet` requires. Without it `pyarrow` doesn't handle pandas metadata at all. The `pyarrow` writer in the recent release of `dask` handles writing this properly, and everything works. This is using `dask` 0.16.0 and `pyarrow` 0.7.1.

```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: import dask.dataframe as dd

In [4]: df = pd.DataFrame(data=np.arange(0,500).reshape(100, 5), columns=['col_A', 'col_B', 'col_C', 'col_D', 'col_E'])
   ...: df.set_index('col_A', inplace=True)
   ...:

In [5]: ddf = dd.from_pandas(df, npartitions=4)

In [6]: ddf.to_parquet('test.parquet', engine='pyarrow')

In [7]: ddf2 = dd.read_parquet('test.parquet', engine='pyarrow')

In [8]: ddf2.head()
Out[8]:
       col_B  col_C  col_D  col_E
col_A
0          1      2      3      4
5          6      7      8      9
10        11     12     13     14
15        16     17     18     19
20        21     22     23     24
```

**Comment by bendruitt** [comment]
Awesome! Thanks for the quick reply. I didn't think that parquet format required a seperate metadata file, so thanks for the correction. I think I'll go forward with dask 0.16.0 and pyarrow 0.7.1 and see if I can use my existing datasets. If that fails, I'll refactor and try using write_parquet function with engine='pyarrow'. Thanks again, and I'll update this thread with the result.

Any comment on the huge size and slow read write of the fastparquet metadata files when used with datasets with high numbers of dimensions?

**Comment by martindurant** [comment]
@bendruitt , on the latter: the metadata will scale as the number of row_groups * the number of columns, because every single chunk has some information saved about it in the metadata. In fastparquet, the metadata reading is faster than it was, and I imagine in arrow it's at least as fast, but you can always have too many columns and rows. It may be worth our considering turning off the saving of extra data in the metadata such as data statistics, which are designed to allow skipping reading some of the data, but perhaps are not very useful in some cases.

**Comment by mrocklin** [comment]
It sounds like much of this has been resolved.  Closing.
